File: training/javascript/src/de/sots/cellarsandcentaurs/domain/service/CreatureService.test.js
import { test } from "node:test";
import assert from "node:assert";
import { CreatureService } from "./CreatureService.js";
import { Creature } from "../model/Creature.js";
import { CreatureId } from "../model/CreatureId.js";
import { Speed } from "../model/Speed.js";

function should_save_creature_to_the_stable() {
    const saved = [];
    const creatures = { save: (creature) => saved.push(creature) };
    const service = new CreatureService(creatures);
    const walking_speed = new Speed(40);
    const creature = new Creature(new CreatureId("centaur-1"));
    creature.setSpeeds(new Map([["walking", walking_speed]]));

    service.save(creature);

    assert.strictEqual(saved.length, 1);
}

test("saves a creature to the stable", should_save_creature_to_the_stable);

File: sample-projects/kotlin/src/test/kotlin/de/sots/cellarsandcentaurs/domain/service/CreatureServiceTest.kt
package de.sots.cellarsandcentaurs.domain.service

import de.sots.cellarsandcentaurs.domain.model.Creature
import de.sots.cellarsandcentaurs.domain.model.CreatureId
import de.sots.cellarsandcentaurs.domain.model.Speed
import de.sots.cellarsandcentaurs.domain.model.SpeedType
import io.mockk.mockk
import io.mockk.verify
import kotlin.test.Test
import kotlin.test.assertEquals

class CreatureServiceTest {

    @Test
    fun should_save_creature_to_the_stable() {
        // Arrange
        val creatures = mockk<Creatures>(relaxed = true)
        val creatureService = CreatureService(creatures)
        val walking_speed = Speed(40)
        val XPValue = 200
        val creature = Creature(CreatureId("centaur-1"))
        creature.speeds = mapOf(SpeedType.WALKING to walking_speed)

        // Act
        creatureService.save(creature)

        // Assert
        verify { creatures.save(creature) }
        assertEquals(40, creature.speedOf(SpeedType.WALKING)?.feetPerRound)
        assertEquals(200, XPValue)
    }
}
